From stable+bounces-274911-greg=kroah.com@vger.kernel.org Wed Jul 15 13:43:49 2026
From: Sasha Levin <sashal@kernel.org>
Date: Wed, 15 Jul 2026 07:40:17 -0400
Subject: io_uring/rw: ensure reissue path is correctly handled for IOPOLL
To: stable@vger.kernel.org
Cc: Jens Axboe <axboe@kernel.dk>, John Garry <john.g.garry@oracle.com>, Sasha Levin <sashal@kernel.org>
Message-ID: <20260715114018.726395-1-sashal@kernel.org>

From: Jens Axboe <axboe@kernel.dk>

[ Upstream commit bcb0fda3c2da9fe4721d3e73d80e778c038e7d27 ]

The IOPOLL path posts CQEs when the io_kiocb is marked as completed,
so it cannot rely on the usual retry that non-IOPOLL requests do for
read/write requests.

If -EAGAIN is received and the request should be retried, go through
the normal completion path and let the normal flush logic catch it and
reissue it, like what is done for !IOPOLL reads or writes.

Fixes: d803d123948f ("io_uring/rw: handle -EAGAIN retry at IO completion time")
Reported-by: John Garry <john.g.garry@oracle.com>
Link: https://lore.kernel.org/io-uring/2b43ccfa-644d-4a09-8f8f-39ad71810f41@oracle.com/
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Stable-dep-of: c554246ff4c6 ("io_uring/rw: preserve partial result for iopoll")
Signed-off-by: Sasha Levin <sashal@kernel.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 io_uring/rw.c |    7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

--- a/io_uring/rw.c
+++ b/io_uring/rw.c
@@ -550,11 +550,10 @@ static void io_complete_rw_iopoll(struct
 	if (kiocb->ki_flags & IOCB_WRITE)
 		io_req_end_write(req);
 	if (unlikely(res != req->cqe.res)) {
-		if (res == -EAGAIN && io_rw_should_reissue(req)) {
+		if (res == -EAGAIN && io_rw_should_reissue(req))
 			req->flags |= REQ_F_REISSUE | REQ_F_BL_NO_RECYCLE;
-			return;
-		}
-		req->cqe.res = res;
+		else
+			req->cqe.res = res;
 	}
 
 	/* order with io_iopoll_complete() checking ->iopoll_completed */
